q2_types_genomics/reference_db/_format.py

```diff
@@ -54,3 +57,84 @@ def _validate_(self, level):
 plugin.register_formats(DiamondDatabaseFileFmt, DiamondDatabaseDirFmt)
 plugin.register_semantic_type_to_format(ReferenceDB[Diamond],
                                         DiamondDatabaseDirFmt)
+
+
+class NCBITaxonomyNodesFormat(model.TextFileFormat):
+    def _validate_n_records(self, n=None):
+        with open(str(self), "r") as fh:
+            file_ = enumerate(fh) if n is None else zip(range(n), fh)
+
+            for i, line in file_:
+                line = line.rstrip("\n").split("\t|\t")
+                if 13 > len(line) or len(line) > 18:
+                    raise ValidationError(
+                        "NCBI taxonomy nodes file must have 13 columns, "
+                        f"found {len(line)} columns on line {i + 1}."
+                    )
+                if not line[0].isnumeric() or not line[1].isnumeric():
+                    raise ValidationError(
+                        "NCBI taxonomy nodes file must contain a numeric "
+                        "taxonomy ID in the first two columns, found "
+                        f"non-numeric value on line {i + 1}."
+                    )
+                for col in (5, 7, 9, 10, 11):
+                    if not line[col].isnumeric() or \
+                            not int(line[col]) in (0, 1):
+                        raise ValidationError(
+                            "NCBI taxonomy nodes file must contain 0 or 1 "
+                            "in columns 6, 8, 10, 11, and 12, found a "
+                            f"non-allowed value on line {i + 1}, column "
+                            f"{col + 1}: {line[col]}."
+                        )
+
+    def _validate_(self, level):
+        self._validate_n_records(n={"min": 10, "max": None}[level])
+
+
+class NCBITaxonomyNamesFormat(model.TextFileFormat):
+    def _validate_n_records(self, n=None):
+        with open(str(self), "r") as fh:
+            file_ = enumerate(fh) if n is None else zip(range(n), fh)
+
+            for i, line in file_:
+                line = line.rstrip("\n").split("\t|\t")
+                if len(line) != 4:
+                    raise ValidationError(
+                        "NCBI taxonomy names file must have 4 columns, "
+                        f"found {len(line)} columns on line {i + 1}."
+                    )
+                if not line[0].isnumeric():
+                    raise ValidationError(
+                        "NCBI taxonomy name file must contain a numeric "
+                        "taxonomy ID in the first column, found non-numeric "
+                        f"value on line {i + 1}: {line[0]}."
+                    )
+
+    def _validate_(self, level):
+        self._validate_n_records(n={"min": 10, "max": None}[level])
+
+
+class NCBITaxonomyBinaryFileFmt(model.TextFileFormat):
+    def _validate_(self, level):
+        pass
```

Comment:
When you un-gzip that file (`prot.accession2taxid.gz`), can you see what the content is? Is it just some mapping of protein accession IDs to taxa IDs? If so, maybe we could provide some sort of validation for that?

Reply:
Indeed it is some sort of mapping between accession IDs and Taxa IDs. It looks like this:

```
accession       accession.version       taxid   gi
A0A009IHW8      A0A009IHW8.1    1310613 1835922267
A0A011QK89      A0A011QK89.1    522306  2203519076
A0A017SE81      A0A017SE81.1    1388766 2316836429
A0A017SE85      A0A017SE85.1    1388766 2316836434
A0A017SEF3      A0A017SEF3.1    1388766 2316836433
A0A017SEX7      A0A017SEX7.1    1388766 2316836426
A0A017SEY2      A0A017SEY2.1    1388766 2316836432
A0A017SFB8      A0A017SFB8.1    1388766 2316836430
A0A017SGC7      A0A017SGC7.1    1388766 2316836431
```

I will validate: 
- that it has the correct columns and 
- validate that both the accession IDs and the Taxa IDs are valid. 

Since the file is 45 Gb (1 billion rows) I will try to validate it on the fly (without decompressing it into a temporary directory).